File: app/src/main/java/com/esoxjem/movieguide/BaseApplication.java
package com.esoxjem.movieguide;

import android.app.Application;
import android.os.StrictMode;

import com.esoxjem.movieguide.details.DetailsComponent;
import com.esoxjem.movieguide.details.DetailsModule;
import com.esoxjem.movieguide.favorites.FavoritesModule;
import com.esoxjem.movieguide.listing.ListingComponent;
import com.esoxjem.movieguide.listing.ListingModule;
import com.esoxjem.movieguide.network.NetworkModule;

import io.realm.Realm;

/**
 * @author arun
 */
public class BaseApplication extends Application {
    private AppComponent appComponent;
    private DetailsComponent detailsComponent;
    private ListingComponent listingComponent;

    @Override
    public void onCreate() {
        super.onCreate();
        StrictMode.enableDefaults();
        initRealm();
        appComponent = createAppComponent();
    }

    private AppComponent createAppComponent() {
        return DaggerAppComponent.builder()
                .appModule(new AppModule(this))
                .networkModule(new NetworkModule())
                .favoritesModule(new FavoritesModule())
                .build();
    }

    private void initRealm() {
        Realm.init(this);
    }

    public DetailsComponent createDetailsComponent() {
        detailsComponent = appComponent.plus(new DetailsModule());
        return detailsComponent;
    }

    public void releaseDetailsComponent() {
        detailsComponent = null;
    }

    public ListingComponent createListingComponent() {
        listingComponent = appComponent.plus(new ListingModule());
        return listingComponent;
    }

    public void releaseListingComponent() {
        listingComponent = null;
    }

    public ListingComponent getListingComponent() {
        return listingComponent;
    }
}

File: app/src/main/java/com/esoxjem/movieguide/BaseApplication.kt
package com.esoxjem.movieguide

import android.app.Application
import android.os.StrictMode

import com.esoxjem.movieguide.movies.details.DetailsComponent
import com.esoxjem.movieguide.movies.details.DetailsModule
import com.esoxjem.movieguide.movies.favorites.FavoritesModule
import com.esoxjem.movieguide.movies.listing.ListingComponent
import com.esoxjem.movieguide.movies.listing.ListingModule
import com.esoxjem.movieguide.network.NetworkModule

/**
 * @author arunsasidharan
 */
class BaseApplication : Application() {
    private lateinit var appComponent: AppComponent
    lateinit var detailsComponent: DetailsComponent
    lateinit var listingComponent: ListingComponent

    override fun onCreate() {
        super.onCreate()
        StrictMode.enableDefaults()
        appComponent = createAppComponent()
    }

    private fun createAppComponent(): AppComponent {
        return DaggerAppComponent.builder()
                .appModule(AppModule(this))
                .networkModule(NetworkModule())
                .favoritesModule(FavoritesModule())
                .build()
    }

    fun createDetailsComponent(): DetailsComponent {
        detailsComponent = appComponent!!.plus(DetailsModule())
        return detailsComponent
    }

    fun createListingComponent(): ListingComponent {
        listingComponent = appComponent!!.plus(ListingModule())
        return listingComponent
    }
}
